### PR TITLE
Add `LRScheduler.state_dict()` to checkpoints

### LINKED ISSUE #2730: Use ``LRScheduler.load_state_dict()``
Currently, when resuming from a previous run that utilizes a learning rate scheduler, we do NOT load a state dict from the scheduler. 

**But wait, does that mean our code is BROKEN?**

Actually, b/c we save the state dict for the optimizer that initializes the learning rate scheduler AND we track the global steps taken which informs the learning rate scheduler where in it's schedule it is, the behavior is largely the same. However, if you were to inspect the learning rate scheduler state dict before and after, they will *not* be the same due to the parameter `_step_count` not being updated. 

**Does this matter?** 

`_step_count` looks like it's mostly there for debugging purposes. It's only *actually* used in one standard PyTorch learning rate scheduler: `CosineAnnealingLR` [here](https://github.com/pytorch/pytorch/blob/134179474539648ba7dee1317959529fbd0e7f89/torch/optim/lr_scheduler.py#L994). I'm opening this issue b/c even though our training code works fine for 99% of use cases, we really should utilize the state dict to cover all the cases.

## Goal

Update our recipes to save the ``lr_scheduler.state_dict()`` to the intermediate state dict and upon resuming from checkpoint, we should call `lr_scheduler.load_state_dict()` on the learning rate scheduler.

### PR DESCRIPTION
**Pull request by omkar-334**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
Resolves #2730

#### Changelog
What are the changes made in this PR?
Adds loading and saving `LRScheduler.state_dict()` from and to checkpoints

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by omkar-334** [comment]
tag @joecummings

**Comment by omkar-334** [comment]
@pbontrager @krammnic @joecummings can you review this?

**Comment by krammnic** [comment]
@omkar-334 On the first glance LGTM. Can you provide the results of the sanity checks? (some evidence of correctness through run)
